llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -3012,7 +3012,7 @@ static Value *simplifyICmpWithConstant(CmpPredicate Pred, Value *LHS,
   }
 
   // Rule out tautological comparisons (eg., ult 0 or uge 0).
-  ConstantRange RHS_CR = ConstantRange::makeExactICmpRegion(Pred, *C);
+  ConstantRange RHS_CR = ConstantRange::makeAllowedICmpRegion(Pred, *C);
```

Comment:
While it doesn't change the outcome, it changes the semantics. We expect `makeExactICmpRegion` to return an exact result instead of a superset.

Can we change `makeExactICmpRegion` to support `samesign`?

Reply:
So, there is a problem with that, because samesign ranges are fundamentally asymmetric; the following assert fails:

```
  assert(makeAllowedICmpRegion(Pred, C) == makeSatisfyingICmpRegion(Pred, C));
```

which leads to the expectation in callers being broken, so that code like this:

```
  const auto CR1 = ConstantRange::makeExactICmpRegion(Pred1, *RHSC1);
  const auto CR2 = ConstantRange::makeExactICmpRegion(Pred2, *RHSC2);

  return CR1.inverse() == CR2;
```

(ValueTracking.cpp:8617)

regresses, even if the assert is removed.

Not sure if there is some better solution to the situation, but my proposal is to have makeAllowedICmpRegion region respect samesign, and makeExactICmpRegion not respect it.

Reply:
Thought about this for a bit, and you're right: it is _semantically_ incorrect. There is no bug because we're passing a singleton value to makeAllowedICmpRegion, but I agree that the situation is bad. I'm not sure how to get out of this jam of fundamentally asymmetric samesign ranges: maybe the best thing we can do is to not touch existing functions and create a makeAsymmetricICmpRegion that accepts a CmpPredicate and ConstantRange, and document that it should not be called with a plain predicate? Not sure what else we can possibly do.